docs/proposals/makefile-improvement-proposal.md

```diff
@@ -0,0 +1,67 @@
+In order to better address reproducibility, it is better to split up the duties the Makefile currently performs into various sub components, that can be assembled in more appropriate technology. One important aspect here is to completely shift the responsibility of dependency management away from the Makefile and into technology such as Nix or Devenv. This proposal will also enable quicker access to a development build of Argo Workflows to developers, reducing the costs of on-boarding and barrier to entry.
+
+### Devenv
+
+#### Benefits of Devenv
+
+- Reproducible build environment
+- Ability to run processes
+
+#### Disadvantages of Devenv
+
+- Huge learning curve to tap into Nix functionality
+- Less documentation
+
+### Nix
+
+#### Benefits of Nix
+
+- Reproducible build environment
+- Direct raw control of various Nix related functionality instead of using Devenv
+- More documentation
+
+#### Disadvantages of Nix
+
+- Huge learning curve
+
+### Recommendation
+
+I suggest that we use Nix over Devenv. I believe that our build environment is unique enough that we will be tapping into Nix anyway, it probably makes sense to directly use Nix in that case.
+
+### Proposal
+
+In order to maximize the benefit we receive from using something like Nix, I suggest that we initially start off with a modest change to the Makefile.
+The first proposal would be to remove out all dependency management code and replace this functionality with Nix, where it is trivially possible. This may not be possible for some go lang related binaries we use, we will retain the Makefile functionality in those cases, at least for a while. Eventually we will migrate more and more of this responsibility away from the Makefile. Following Nix being responsible for all dependency management, we could start to consider moving more of our build system itself into Nix, perhaps it is easiest to start off with UI build as it is relatively painless. However, do note that this is not a requirement, I do not see a problem with the Makefile and the Nix file co-existing, it is more about finding a good balance between the reproducibility we desire and the effort we put into obtaining said reproducibility. An example for a replacement could be [this](https://github.com/argoproj/argo-workflows/blob/master/Makefile#L629) dependency for example, note that we do not state any version here, replacing such installations with Nix based installations will ensure that we can ensure that if a build works on a certain developer's machine, it should also work on every other machine as well.
+
+### What will Nix get us?
+
+As mentioned previously Nix gets us closer to reproducible build environments. It should ease significantly the on-boarding process of developers onto the project.
+There have been several developers who wanted to work on Argo Workflows but found the Makefile to be a barrier, it is likely that there are more developers on this boat. With a reproducible build environment, we hope that
```

Comment:
Can you clarify some of the problems faced? And were they Go modules or React/JS/TS?

Reply:
Well I run a Fedora machine, the Makefile just isn't suited for this. 
When I initially on boarded to this project,  I had to spend a week or so getting all the dependencies exactly right. 
I went on leave and came back only to have to spend another two days to get Argo Workflows working again. 

In fact, the Makefile is broken on my laptop as we speak. My development process when working from my laptop is to push changes, pull them to my desktop and check everything on there. 

`make pre-commit` pretty much never works for me.

I think providing a Hermetic build system is a massive improvement over this. 
With my latest PR, I was able to get Argo running in < 15 minutes. This was a huge improvement.

Reply:
I do not believe having global installs of packages like [this](https://github.com/argoproj/argo-workflows/blob/master/Makefile#L640) are great for reproducibility. 
For example it might be entirely feasible for some pip packages to not work on certain versions of Python. 

Argo relies on a lot of external tooling, I think it would be a good idea to pin these down to exact versions. 
As an example [these](https://github.com/argoproj/argo-workflows/pull/10999/files#diff-206b9ce276ab5971a2489d75eb1b12999d4bf3843b7988cbe8d687cfde61dea0R37-R141) download exactly a known working combination of python interpreter and packages.